compiler+runtime/src/cpp/jank/runtime/obj/ratio.cpp

```diff
@@ -0,0 +1,556 @@
+#include <jank/runtime/obj/ratio.hpp>
+#include <fmt/compile.h>
+#include <limits>
+
+namespace jank::runtime
+{
+
+  constexpr native_real EPSILON = std::numeric_limits<native_real>::epsilon();
+
+  obj::ratio_data::static_object(native_integer numerator, native_integer denominator)
+    : numerator{ numerator }
+    , denominator{ denominator }
+  {
+    if(denominator == 0)
+    {
+      throw std::invalid_argument("Ratio denominator cannot be zero.");
+    }
+    simplify();
+  }
+
+  object_ptr obj::ratio::create(native_integer const numerator, native_integer const denominator)
+  {
+    auto ratio_data = make_box<obj::ratio_data>(numerator, denominator);
+    if(ratio_data->denominator == 1)
+    {
+      return make_box<obj::integer>(ratio_data->numerator);
+    }
+    auto r(make_box<obj::ratio>());
+    r->data.numerator = ratio_data->numerator;
+    r->data.denominator = ratio_data->denominator;
+    return r;
+  }
+
+  native_real obj::ratio_data::to_real() const
+  {
+    return static_cast<native_real>(numerator) / denominator;
+  }
+
+  native_real obj::ratio::to_real() const
+  {
+    return data.to_real();
+  }
+
+  native_integer obj::ratio::to_integer() const
+  {
+    return data.numerator / data.denominator;
+  }
+
+  void obj::ratio_data::simplify()
+  {
+    int gcd = std::gcd(numerator, denominator);
+    numerator /= gcd;
+    denominator /= gcd;
+
+    if(denominator < 0)
+    {
+      numerator = -1 * numerator;
+      denominator = -1 * denominator;
+    }
+  }
+
+  void obj::ratio::to_string(fmt::memory_buffer &buff) const
+  {
+    format_to(std::back_inserter(buff), FMT_COMPILE("{}/{}"), data.numerator, data.denominator);
+  }
+
+  native_persistent_string obj::ratio::to_string() const
+  {
+    fmt::memory_buffer buff;
+    to_string(buff);
+    return native_persistent_string{ buff.data(), buff.size() };
+  }
+
+  native_persistent_string obj::ratio::to_code_string() const
+  {
+    return to_string();
+  }
+
+  native_hash obj::ratio::to_hash() const
+  {
+    return hash::combine(hash::integer(data.numerator), hash::integer(data.denominator));
+  }
+
+  native_bool obj::ratio::equal(object const &o) const
+  {
+    if(o.type == object_type::integer)
+    {
+      auto const i(expect_object<obj::integer>(&o));
+      return this->data == i->data;
+    }
+
+    if(o.type == object_type::real)
+    {
```

Comment:
Rather than checking the type and then using `expect_object` as a pattern, please use `dyn_cast`.

```cpp
    if(auto const i = dyn_cast<obj::integer>(&o))
    {
      return this->data == i->data;
    }
```

Note that C++ doesn't allow direct initialization in this one case of initializing within an `if`.

Also, there is no need for `this->` here, so please remove that.

Reply:
updated. the reason for that is, we don't want code to terminate in case `expect_object` fails?